lib/core/CLoggerThrottler.cc

```diff
@@ -0,0 +1,45 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+#include <core/CLoggerThrottler.h>
+
+#include <core/CTimeUtils.h>
+
+#include <limits>
+
+namespace ml {
+namespace core {
+
+CLoggerThrottler::CLoggerThrottler() : m_MinimumLogIntervalMs{3600 * 1000} {
+}
+
+void CLoggerThrottler::minimumLogIntervalMs(std::int64_t minimumLogIntervalMs) {
+    m_MinimumLogIntervalMs = minimumLogIntervalMs;
+}
+
+std::pair<std::size_t, bool> CLoggerThrottler::skip(const char* file, int line) {
+
+    auto key = std::make_pair(file, line);
```

Comment:
One interesting question about using `__FILE__` as part of the key here is whether it's the same if a log message is in a header file and that header file is included in different implementation files.  For example, `CLinearAlgebra.h` contains a `LOG_ERROR` macro, so is the `__FILE__` for `CLinearAlgebra.h` the same in both `CSeasonalComponentAdaptiveBucketing.cc` and `CTimeSeriesDecompositionDetail.cc`?  The answer might vary between compilers too.

If it does vary I don't think that's a big enough problem to warrant changing to a less efficient implementation, but it would deserve a comment to say it's a known problem and sufficiently minor to be acceptable.

Reply:
Thinking about logging I've seen in the passed I think `__FILE__` nearly always resolves to the header name. However, I added a warning. There is also a work around that the line is explicitly wrapped in a check supplying a custom unique string. I think this is another good reason for not trying to do any more complex association of log messages. See [80fc1ca](https://github.com/elastic/ml-cpp/pull/1615/commits/80fc1cad6711f5fbf6050c173a87ce8657934487).

Reply:
> `__FILE__` nearly always resolves to the header name

Yes, it does, but you're keying on a pointer to the header name, which might well vary between compiler invocations.

Reply:
I'd also wondered this, and in fairness I didn't try other compilers, but clang certainly uses a unique string literal. I think it would be poor compiler behaviour to duplicate strings in this fashion as it would lead to unnecessary executable bloat.